### PR TITLE
feat: add initial tracing to transaction controller

### PR DESCRIPTION
**Pull request by matthewwalsh0**
## Explanation

Add an optional `trace` constructor callback to the `TransactionController` to support generating Sentry traces within the controller, without coupling to the Sentry package.

Add an optional `traceContext` argument to `addTransaction` to facilitate generating nested traces within a parent trace generated within a client.

Add initial instrumentation of the transaction lifecycle.

## References

[#2849](https://github.com/MetaMask/MetaMask-planning/issues/2849) [#2903](https://github.com/MetaMask/MetaMask-planning/issues/2903)

### `@metamask/transaction-controller`

- **ADDED**: Add optional `trace` callback to constructor.
- **ADDED**: Add optional `traceContext` option to `addTransaction` method.

## Checklist

- [x] I've updated the test suite for new or updated code as appropriate
- [x] I've updated documentation (JSDoc, Markdown, etc.) for new or updated code as appropriate
- [x] I've highlighted breaking changes using the "BREAKING" category above as appropriate

### CODE REVIEW

**Comment by matthewwalsh0** [comment]
@metamaskbot publish-preview

**Comment by matthewwalsh0** [comment]
@metamaskbot publish-preview

File: packages/transaction-controller/src/types.ts [hunk 1]
```diff
@@ -1299,3 +1299,40 @@ export type SimulationData = {
   /** Data concerning a change to the user's token balances. */
   tokenBalanceChanges: SimulationTokenBalanceChange[];
 };
+
+/** A context in which to execute a trace, in order to generate nested timings. */
+export type TraceContext = unknown;
```
Comment (matthewwalsh0) [inline packages/transaction-controller/src/types.ts:1304]: If this proves successful, we could migrate these to the `controller-utils`.

**Review by pedronfigueiredo — APPROVED** [review]

**Comment by matthewwalsh0** [comment]
@metamaskbot publish-preview

**Pull request merged**